pandas/_libs/algos.pyx

```diff
@@ -1231,15 +1237,31 @@ def diff_2d(
                         start, stop = 0, sx + periods
                     for j in range(sy):
                         for i in range(start, stop):
-                            out[i, j] = arr[i, j] - arr[i - periods, j]
+                            left = arr[i, j]
+                            right = arr[i - periods, j]
+                            if out_t is int64_t and datetimelike:
```

Comment:
you could make this simpler by
```
if out_t is int64_t and datetimelike and (left == NPY_NAT or right == NPY_NAT):
   .....
else
   ...
```
to make slightly shorter, not a big deal though

Reply:
i find this clearer